library/public/window.h

```diff
@@ -0,0 +1,98 @@
+/**
+ * @class   window
+ * @brief   Abstract class to render in a window or an image
+ *
+ * A class to render things in a window or an image.
+ * An icon and windowName can be set which can be shown by a window manager.
+ */
+
+#ifndef f3d_window_h
+#define f3d_window_h
+
+#include "camera.h"
+#include "export.h"
+#include "image.h"
+
+#include <string>
+
+namespace f3d
+{
+class F3D_EXPORT window
+{
+public:
+  class exception : public std::runtime_error
+  {
+  public:
+    exception(const std::string& what = "")
+      : std::runtime_error(what)
+    {
+    }
+  };
+
+  /**
+   * Enumeration of supported window types
+   * =====================================
+   * NONE: A mock window without rendering capabilities
+   * NATIVE: A window using the native graphical stack
+   * NATIVE_OFFSCREEN: A native window rendering to an offscreen buffer, not visible on screen
+   * EXTERNAL: An external window that assume the OpenGL context would have been created by
+   * another framework
+   */
+  enum class Type : unsigned char
+  {
+    NONE,
+    NATIVE,
+    NATIVE_OFFSCREEN,
+    EXTERNAL
+  };
+
+  /**
+   * Get the type of the window
+   */
+  virtual Type getType() = 0;
+
+  /**
+   * Get the camera provided by the window.
+   */
+  virtual camera& getCamera() = 0;
+
+  /**
+   * Perform a render of the window to the screen.
+   * All dynamic options are updated if needed.
+   * Return true on success, false otherwise.
+   */
+  virtual bool render() = 0;
+
+  /**
+   * Perform a render of the window to the screen and save the result in a f3d::image.
+   * Set noBackground to true to have a transparent background.
+   * Return the resulting f3d::image.
+   */
+  virtual image renderToImage(bool noBackground = false) = 0;
+
+  /**
+   * Set the size of the window
+   */
+  virtual window& setSize(int width, int heigth) = 0;
+
+  /**
+   * Set the icon to be shown by a window manager.
+   * icon should be an unsigned char array.
+   * iconSize should be the sizeof(icon).
+   */
+  virtual window& setIcon(const void* icon, size_t iconSize) = 0;
```

Comment:
If the icon should be an unsigned char array, why not actually require a `const unsigned char*`?

Reply:
I don't know what your policy on requiring other libraries is, but consider something like `gsl::span` from the guidelines support library here.

https://github.com/gsl-lite/gsl-lite
https://github.com/gsl-lite/gsl-lite/blob/4720a2980a30da085b4ddb4a0ea2a71af7351a48/include/gsl/gsl-lite.hpp#L3649

Reply:
I agree about the type of the pointer.  
For gsl, we should avoid it, especially in the public headers.

Reply:
indeed

Reply:
Adressed in https://github.com/f3d-app/f3d/pull/355